lib/travis/hub/api/auth.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+require 'rack/auth/abstract/handler'
+require 'rack/auth/abstract/request'
+
+module Travis
+  module Hub
+    class Auth < Rack::Auth::AbstractHandler
+      attr_reader :tokens
+
+      def initialize(app, config = {}, alg = 'RS512')
+        @app = app
+        @alg = alg
+        @key = OpenSSL::PKey::RSA.new(config[:jwt_public_key])
+        @tokens = config[:http_basic_auth]
+        @verify = true
+      end
+
+      attr_reader :alg, :key, :verify
+
+      def call(env)
+        auth = Request.new(env)
+
+        return unauthorized unless auth.provided?
+        return bad_request  unless auth.basic? || auth.bearer?
+
+        if basic_valid?(auth)
+          env['REMOTE_USER'] = auth.basic_username
+          return @app.call(env)
+        end
+
+        unless decode_jwt(auth)
+          return [403, { 'Content-Type' => 'text/plain', 'Content-Length' => '0' }, []]
+        end
+
+        if bearer_valid?(auth)
+          # env['jwt.header'] = auth.jwt_header
+          # env['jwt.payload'] = auth.jwt_payload
+          return @app.call(env)
+        end
+
+        unauthorized
+      end
+
+      private
+
+      def challenge
+        'Basic realm="job-board"'
+      end
+
+      def basic_valid?(auth)
+        user, password = auth.basic_credentials
+        return unless user && password
+        tokens[user.to_sym] == password
+      end
+
+      def bearer_valid?(auth)
+        return false if auth.job_id.nil?
```

Comment:
@BanzaiMan i'd normally agree, but there's now that other (previously commented out, now commented back in) condition that would make it a very long line, so i think i'll keep the current style?